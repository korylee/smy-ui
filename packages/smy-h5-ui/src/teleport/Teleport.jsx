import { getSlot } from '../_utils/vue/slots'
import Vue from 'vue'
import { props } from './props'

export default {
  name: 'SmyTeleport',
  props,
  data: () => ({
    el: null,
    instance: null,
  }),
  watch: {
    disabled: 'transfer',
    to: 'transfer',
  },
  methods: {
    create() {
      const Ctor = Vue.extend({
        render: () => <div class="smy-teleport__container">{getSlot(this)}</div>,
      })
      this.instance = new Ctor()
      this.instance.$parent = this
      this.el = this.instance.$mount().$el
    },
    transfer() {
      const container = this.disabled ? this.$refs.teleport : document.querySelector(this.to)
      const parentNode = this.el.parentNode
      if (parentNode === container) return
      parentNode?.removeChild(this.el)
      container.appendChild(this.el)
    },
  },
  updated() {
    this.instance.$forceUpdate()
  },
  mounted() {
    this.create()
    this.transfer()
  },
  deactivated() {
    this.el.parentNode?.removeChild(this.el)
  },
  activated() {
    this.transfer()
  },
  beforeDestroy() {
    this.instance.$destroy()
    this.el.parentNode?.removeChild(this.el)
  },

  render() {
    return <div ref="teleport" class="smy-teleport" />
  },
}
