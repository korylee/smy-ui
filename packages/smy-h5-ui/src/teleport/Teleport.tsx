import { SlotsMixin } from '@smy-h5/vtools'
import Vue from 'vue'
import { props } from './props'

export default {
  name: 'SmyTeleport',
  mixins: [SlotsMixin],
  props,
  data: () => ({
    el: null,
    instance: null,
  }),
  methods: {
    create() {
      const Ctor = Vue.extend({
        render: () => <div class="smy-teleport__container">{this.getSlots()}</div>,
      })
      this.instance = new Ctor()
      this.instance.$parent = this
      this.el = this.instance.$mount().$el
    },
    transfer() {
      const container = this.disabled ? this.$refs.teleport : document.querySelector(this.to)
      const parentNode = this.el.parentNode
      if (parentNode === container) return
      parentNode?.removeChild(this.el)
      container.appendChild(this.el)
    },
  },
  updated() {
    this.instance.$forceUpdate()
  },
  mounted() {
    this.create()
    this.transfer()
  },
  deactivated() {
    this.el.parentNode?.removeChild(this.el)
  },
  activated() {
    this.transfer()
  },
  beforeDestroy() {
    this.instance.$destroy()
  },

  render() {
    return <div ref="teleport" class="smy-teleport" />
  },
}
