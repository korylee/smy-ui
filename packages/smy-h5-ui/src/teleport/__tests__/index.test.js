import { mount } from '@vue/test-utils'
import Vue from 'vue'
import Teleport from '..'

test('test countdown plugin', () => {
  Vue.use(Teleport)
  expect(Vue.component(Teleport.name)).toBeTruthy()
})

const Wrapper = {
  components: {
    SmyTeleport: Teleport,
  },
  data: () => ({ show: true, disabled: false, keepAlive: false, to: 'body' }),
  template: `
    <component :is="keepAlive ? 'keep-alive' : 'div'">
      <smy-teleport v-if="show" :to="to" :disabled="disabled">hello world</smy-teleport>
    </component>
  `,
}

test('test teleport transfer', () => {
  const wrapper = mount(Wrapper)
  expect(wrapper.html()).toMatchSnapshot()
  expect(document.body.innerHTML).toBe('<div class="smy-teleport__container">hello world</div>')

  wrapper.destroy()
  expect(document.body.innerHTML).toBe('')
})

test('test teleport disabled', async () => {
  const wrapper = mount(Wrapper)
  expect(document.body.innerHTML).toBe('<div class="smy-teleport__container">hello world</div>')

  await wrapper.setData({ disabled: true })
  expect(document.body.innerHTML).toBe('')
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.destroy()
})

test('test teleport with keepalive', async () => {
  const wrapper = mount(Wrapper)

  await wrapper.setData({ keepAlive: true })
  expect(document.body.innerHTML).toBe('<div class="smy-teleport__container">hello world</div>')

  await wrapper.setData({ show: false })
  expect(document.body.innerHTML).toBe('')

  await wrapper.setData({ show: true })
  expect(document.body.innerHTML).toBe('<div class="smy-teleport__container">hello world</div>')
  wrapper.destroy()
})

test('test teleport to', async () => {
  document.body.innerHTML = '<div id="app"></div>'
  const wrapper = mount(Wrapper)
  expect(document.body.innerHTML).toMatchSnapshot()

  await wrapper.setData({ to: '#app' })
  expect(document.body.innerHTML).toMatchSnapshot()

  wrapper.destroy()
})
