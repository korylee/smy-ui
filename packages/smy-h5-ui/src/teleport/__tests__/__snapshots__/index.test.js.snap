// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test teleport disabled 1`] = `
"<div>
  <div class="smy-teleport">
    <div class="smy-teleport__container">hello world</div>
  </div>
</div>"
`;

exports[`test teleport to 1`] = `"<div id="app"></div><div class="smy-teleport__container">hello world</div>"`;

exports[`test teleport to 2`] = `"<div id="app"><div class="smy-teleport__container">hello world</div></div>"`;

exports[`test teleport transfer 1`] = `
"<div>
  <div class="smy-teleport"></div>
</div>"
`;
